### PR TITLE
add XPU support for autocast

### PR DESCRIPTION
**Pull request by guoyejun**

### CODE REVIEW

**Comment by guoyejun** [comment]
there's an error in CI at https://pipelines.actions.githubusercontent.com/serviceHosts/7d146c05-69c3-4c20-a0e7-818111670117/_apis/pipelines/1/runs/1933542/signedlogcontent/239?urlExpires=2022-04-05T10%3A05%3A21.3824994Z&urlSigningMethod=HMACV1&urlSignature=DeTPkfhhK9QMOLNx5Whf6cw%2BlkjGyk46BXW4WAhBsR4%3D, could you share a bit more about this issue? thanks. (this PR does not change bwd of quantized::softmax)

```
2022-04-05T09:57:34.5455965Z processing existing schema:  quantized::softmax(Tensor qx, int dim, float output_scale, int output_zero_point) -> (Tensor)
2022-04-05T09:57:34.5456714Z Can NOT find backward compatible schemas after changes for schema quantized::softmax(Tensor qx, int dim, float output_scale, int output_zero_point) -> (Tensor) from the following candidates:
...
2022-04-05T09:57:34.6509617Z processing existing schema:  enable(__torch__.torch.classes.profiling._ScriptProfile _0) -> (NoneType _0)
2022-04-05T09:57:34.6510742Z processing existing schema:  disable(__torch__.torch.classes.profiling._ScriptProfile _0) -> (NoneType _0)
2022-04-05T09:57:34.6512350Z processing existing schema:  _dump_stats(__torch__.torch.classes.profiling._ScriptProfile _0) -> (__torch__.torch.classes.profiling.SourceStats[] _0)
2022-04-05T09:57:34.6513768Z processing existing schema:  __init__(__torch__.torch.classes.dist_rpc.WorkerInfo _0, str _1, int _2) -> (NoneType _0)
2022-04-05T09:57:34.6514004Z The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 
2022-04-05T09:57:34.6514094Z 
2022-04-05T09:57:34.6514171Z Broken ops: [
2022-04-05T09:57:34.6514414Z 	quantized::softmax(Tensor qx, int dim, float output_scale, int output_zero_point) -> (Tensor)
2022-04-05T09:57:34.6514462Z ]
2022-04-05T09:57:34.7435887Z + cleanup
2022-04-05T09:57:34.7436203Z + retcode=1
2022-04-05T09:57:34.7436514Z + set +x
2022-04-05T09:57:34.7474619Z ##[error]Process completed with exit code 1.
```

File: c10/core/DispatchKey.h [hunk 1]
```diff
@@ -325,6 +325,7 @@ enum class DispatchKey : uint16_t {
   // Autocasting precedes VariableTypeId, to ensure casts are autograd-exposed
   // and inputs are saved for backward in the post-autocast type.
   AutocastCPU,
+  AutocastXPU,
```
Comment (bdhirsh) [inline c10/core/DispatchKey.h:328]: One thing I'm wondering - does XPU really need it's own set of custom autocast kernels? (so the rules are totally different vs the existing cpu / cuda rules?)

@albanD do you know who would be a good person to review extensions to AMP?

Comment (gujinghui) [reply c10/core/DispatchKey.h:328]: @bdhirsh  Thanks for comments.
We will have similar solution for autocast, comparing with CPU and CUDA.
But the detail implementation is not identical.
1. Since CPU/CUDA is almost always available in PyTorch release package, perhaps we cannot reuse or override their dispatch key.
2. The backend capacity is different between AutocastXPU and other device. XPU will have different kernel list, comparing with CPU and CUDA.
3. XPU is out-of-tree device. Each XPU implementation will provide different kernel list, which depends on separate AutocastXPU key.

**Comment by gujinghui** [comment]
@ezyang   @bdhirsh   @albanD 
Could you help review this PR? Thanks a lot.

File: torch/csrc/autograd/init.cpp [hunk 2]
```diff
@@ -574,6 +613,10 @@ static PyMethodDef methods[] = { // NOLINT
   {"get_autocast_cpu_dtype", get_autocast_cpu_dtype, METH_NOARGS, nullptr},
   {"set_autocast_gpu_dtype", set_autocast_gpu_dtype, METH_O, nullptr},
   {"get_autocast_gpu_dtype", get_autocast_gpu_dtype, METH_NOARGS, nullptr},
+  {"set_autocast_xpu_enabled", set_autocast_xpu_enabled, METH_O, nullptr},
+  {"is_autocast_xpu_enabled", is_autocast_xpu_enabled, METH_NOARGS, nullptr},
+  {"set_autocast_xpu_dtype", set_autocast_xpu_dtype, METH_O, nullptr},
+  {"get_autocast_xpu_dtype", get_autocast_xpu_dtype, METH_NOARGS, nullptr},
```
Comment (bdhirsh) [inline torch/csrc/autograd/init.cpp:619]: Do these public python API's need to be in core? I'm wondering if they should come with the xpu extension, given that the actual XPU autocast kernels live with the extension too.

**Comment by bdhirsh** [comment]
cc @ngimel

**Comment by ngimel** [comment]
Autocast bits looks Ok to me inasmuch they are the same as existing cpu and cuda counterparts, but like @bdhirsh I'm not sure if xpu-specific functions should be publicly exposed in core, and I'll leave the decision of what to do with dispatch key to @bdhirsh also. 
TBH, I think for out of core prototype projects like these I think tensor subclassing is the correct solution, it will allow you to define precise op-by-op casting behavior that you want, see here for more info on subclassing and usage examples https://dev-discuss.pytorch.org/t/what-and-why-is-torch-dispatch/557

File: torch/csrc/autograd/init.cpp (hunk 2 above)
Comment (guoyejun) [reply torch/csrc/autograd/init.cpp:619]: sure, let me remove the public python APIs here.

**Comment by guoyejun** [comment]
> Autocast bits looks Ok to me inasmuch they are the same as existing cpu and cuda counterparts, but like @bdhirsh I'm not sure if xpu-specific functions should be publicly exposed in core, and I'll leave the decision of what to do with dispatch key to @bdhirsh also. TBH, I think for out of core prototype projects like these I think tensor subclassing is the correct solution, it will allow you to define precise op-by-op casting behavior that you want, see here for more info on subclassing and usage examples https://dev-discuss.pytorch.org/t/what-and-why-is-torch-dispatch/557

thanks, i'll take a look at this solution, and autocast is also the feature that we'd like to align with stock pytorch.

**Review by bdhirsh — APPROVED** [review]
This seems ok to me, given that we're not using up extra dispatch keys (well we are actually, but we should just convert `Autocast` to be a per-backend-functionality key so that it only takes up one slot).

I would kinda prefer it if there weren't xpu-specific autocast API's exposed, but I don't see another way around it if XPU needs its own version of `autocast_dtype`. I do think that long-term in a world with open dispatch key registration, we would get rid of `get_autocast_xpu_dtype`/`set_autocast_xpu_dtype` in favor or something like `get_autocast_backend_dtype`, but I think it's ok to solve that problem later.
